### PR TITLE
Fix: Create new texture when video frame size is changed.

### PR DESCRIPTION
**Pull request by alexmercerind**
[WIP]

- Add `VideoResizeCallback` to `Player::OnVideo` listener, to get callback ahead of buffer reallocation (if the frame size is changed).
- Now PlayerRegisterTexture & PlayerUnregisterTexture are renamed to PlayerCreateVideoOutlet & PlayerDeleteVideoOutlet.
- Now method channel call no longer creates texture ahead of time, but when first frame of the video is received.
- If a new media of different video dimensions is opened, the previous texture is deleted & new is created (`VideoOutlet::CreateNewTexture`), whose texture ID is notified through the method channel.
- Currently `UnregisterTexture` is causing exception (from `VideoOutlet::DeleteTexture`).
  - Not sure about the reason.
  - Although I'm sure that I'm NOT deleting the texture ID which was never created.

Screenshot 2021-08-25 171022

### CODE REVIEW

File: windows/video_outlet.cc [hunk 1]
```diff
@@ -1,24 +1,50 @@
...
+  return texture_id_.value();
+}
+
+void VideoOutlet::DeleteTexture() {
+  // Prevents calling `UnregisterTexture` if no texture was ever registered.
+  if (texture_id_.has_value()) {
+    channel_->InvokeMethod(
+        "PlayerTextureId",
+        std::make_unique<flutter::EncodableValue>(flutter::EncodableMap(
+            {{"playerId", player_id_}, {"textureId", nullptr}})));
```
Comment (alexmercerind) [inline windows/video_outlet.cc:28]: I decided to just change `textureId` `ValueNotifier<int?>` to `null` before we unregister previous texture & register new one.

**Comment by jnschulze** [comment]
Does the old texture get unregistered before you reallocate the pixel buffer here?https://github.com/alexmercerind/dart_vlc/blob/09805e4ab774ecdc4f169391df47851a48fc9f48/dartvlc/internal/events.h#L146

That's the critical part. If you reuse the same buffer with the new texture you gain nothing.

File: windows/video_outlet.cc [hunk 2]
```diff
@@ -1,24 +1,50 @@
...
+    texture_registrar_->UnregisterTexture(texture_id_.value());
+    texture_id_ = std::nullopt;
+  }
 }
 
 void VideoOutlet::OnVideo(uint8_t* buffer, int32_t width, int32_t height) {
+  if (width_ != width && height_ != height) {
```
Comment (jnschulze) [inline windows/video_outlet.cc:35]: You obviously want to recreate the texture if EITHER the width OR the height changes.

**Comment by alexmercerind** [comment]
I added https://github.com/alexmercerind/dart_vlc/blob/09805e4ab774ecdc4f169391df47851a48fc9f48/dartvlc/internal/events.h#L143-L145 https://github.com/alexmercerind/dart_vlc/blob/09805e4ab774ecdc4f169391df47851a48fc9f48/windows/dart_vlc_plugin.cc#L81-L85 this callback to call `VideoOutlet::DeleteTexture` before `video_frame_buffer_.reset(new uint8_t[size]);`.

And I can confirm that a new `texture_id_` is created before deletion.

> Textures must be unregistered on the the platform thread.

Maybe this is the cause. Maybe I have to involve Dart calling `DeleteTexture` & this cannot be done by VLC thread imo.

**Comment by jnschulze** [comment]
> Textures must be unregistered on the the platform thread.

Where did you get this from?

All texture operations in the texture registrar are thread safe:
https://github.com/flutter/engine/blob/58b1b608c83b624146dc66d98172cac43edaad55/shell/platform/windows/flutter_windows_texture_registrar.cc#L48-L62

File: windows/video_outlet.cc (hunk 2 above)
Comment (alexmercerind) [reply windows/video_outlet.cc:35]: silly mistake. 😅

**Comment by alexmercerind** [comment]
All I need to do is unmount `Texture` before calling `UnregisterTexture`. Let's see...

**Comment by jnschulze** [comment]
In the unregistration path, the mutex only protects the map. The actual texture unregistration is performed asynchronously after the texture has been removed from the map. So even after `TextureRegistrar::UnregisterTexture` returns it's not safe to delete the pixel buffer immediately, as it might still be read by `glTexImage2D` in `shell/platform/windows/external_texture_gl.cc`.

**Comment by alexmercerind** [comment]
Possibly the best would be to publish today's morning patch (that's critical, idk how old `OnVideo` was not working sometimes) & change default value of `videoDimensions` in `Player` constructor to `const VideoDimensions(1920, 1080)`, preventing buffer reallocation until this is fixed.

**Comment by jnschulze** [comment]
> Possibly the best would be to publish today's morning patch (that's critical, idk how old `OnVideo` was not working sometimes) & change default value of `videoDimensions` in `Player` constructor to `const VideoDimensions(1920, 1080)`, preventing buffer reallocation until this is fixed.

I wouldn't change anything on the Dart side. The buffer size doesn't need to match the visible video resolution. You can allocate a fixed buffer of 4096x4096 etc. and don't reallocate it. It shouldn't require any further changes. Having a larger buffer wouldn't even affect performance, as only the visible portion will get copied (both from VLC into the buffer and from the buffer to the GL_TEXTURE_2D used by Flutter).

**Comment by alexmercerind** [comment]
@jnschulze 

> I wouldn't change anything on the Dart side.

Changing C++ code vs temporarily adding a default argument in Dart.
And, users will still have ability to override the frame size per instance of `Player`.

> Having a larger buffer wouldn't even affect performance, as only the visible portion will get copied.

Having high buffer size has actually caused CPU usage to be higher in my tests. 

And, I don't know but I actually tested out different frame sizes when adding dynamic buffer allocation. I'm just unaware how this is coming into sight now... Anyways, it's a problem for now.

**Comment by jnschulze** [comment]
> > Having a larger buffer wouldn't even affect performance, as only the visible portion will get copied.
> 
> Having high buffer size has actually caused CPU usage to be higher in my tests.

Exactly, because you've changed it for the `Player`, right? By doing that, you not only increase the buffer size, but make all buffer-related paths in C++ deal with your forced `width` and `height`, right?

**Comment by alexmercerind** [comment]
> You might want to take a look at your implementation 😄

Literally only `PlayerEvents::OnVideoPictureCallback` out of that is called per frame, remaining is just invoked when media starts for the first time.
I believe it's internal libVLC computation cost that depends on buffer size. They were pretty concerned about using video callbacks for frame in the first place. libVLC 4.0.0 will come with direct rendering to OpenGL & zero copy callbacks, for now this is all there is.

**Comment by jnschulze** [comment]
> > You might want to take a look at your implementation 😄
> 
> Literally only `PlayerEvents::OnVideoPictureCallback` out of that is called per frame, remaining is just invoked when media starts for the first time.
> I believe it's internal libVLC computation cost that depends on buffer size. They were pretty concerned about using video callbacks for frame in the first place. libVLC 4.0.0 will come with direct rendering to OpenGL & zero copy callbacks, for now this is all there is.

Hitesh. VLC doesn't know about your buffer size. Do you tell it the size somewhere? Nope. You tell it the `width` and `height`. So VLC expects you to have a buffer that is at LEAST `width*height*4` for RGBA. But it doesn't prevent you from having a larger buffer. A larger buffer doesn't implicitly affect performance. What affects performance here is that you report a potentially larger `width` and `height` than the actual video has to VLC.

**Comment by alexmercerind** [comment]
I obviously meant the size that is being copied. The larger the dimensions, the larger the size being copied, and in our case that size is equal to the buffer itself. 

I didn't think about pre-defining a large buffer ahead of time & how I can avoid reinitialization. Apologies.

**Comment by jnschulze** [comment]
It's ok.
And as I don't know what the side effects of simple forcing an arbitrary resolution for the Player the Dart side are, I thought it might be simpler to just allocate a buffer that is sufficiently sized.

**Pull request closed without merging**